### PR TITLE
add options to require this/Me for fields, properties, methods, and events

### LINKED ISSUE #7065: Code Style: Analyzer + Code fix for requiring/prohibiting this/Me
We have an analyzer + code fix to support removing redundant uses, but we need options for what type of access it is (field/property/method/etc), and to require it.

### PR DESCRIPTION
**Pull request by brettfo**
### Description
This PR splits the option "Qualify member access with '`this`' or '`Me`'" into four separate options (and marks the former as `[Obsolete]`):

- Qualify member **field** access with '`this`' or '`Me`'.

- Qualify member **property** access with '`this`' or '`Me`'.

- Qualify member **method** access with '`this`' or '`Me`'.

- Qualify member **event** access with '`this`' or '`Me`'.

### Implementation/Rationale
I opted to make all changes in the reducer, including potentially expanding `someField` to `this.someField`/`Me.someField`.  My reasons for doing this are twofold:
1.  The reducer already has to handle the reduction or non-reduction of `this.`/`Me.` and it seemed to make sense to do this work in the same location.
2.  It is a well-known pattern that third-party code fixers should add the `Simplifier.Annotation` annotation to their generated code but they won't necessarily expand (adding `this.`/`Me.`) before adding the annotation, and we want all code fixers (first- and third-party) to have the same nice experience of "I add the `Simplifier.Annotation` annotation and the user's code gets automatically corrected to their specified style."

### Errata
The vast majority of the tests added are in `TypeNameSimplifierTest.vb` but for whatever reason, git classifies this as a huge change, when in reality I just updated 2 tests, then added a bunch of others.  To see the added tests, view the raw file and look for all tests with the `"Qualify"` prefix (e.g., `QualifyFieldAccessWithThis_AsLHS_CSharp`).  They were added in two batches; near the middle of the file for C# and near the end of the file for VB.  There are about 10 or 12 tests for each language.

Most of this was simple plumbing work with the most significant changes occurring in `ExpressionSyntaxExtensions.cs/.vb`.  The changes in `CSharpNameReducer.Rewriter.cs` and `VisualBasicNameReducer.Rewriter.vb` are purely refactorings to make the file shorter and (I think) easier to read.  E.g., the C# file changed a bunch of methods that looked like this:

``` C#
VisitAliasQualifiedName()
{
    bool oldAlwaysSimplify = this.alwaysSimplify;  
    if (!this.alwaysSimplify)  
    {  
        this.alwaysSimplify = node.HasAnnotation(Simplifier.Annotation);  
    }  

    var result = SimplifyExpression(  
        node,  
        newNode: base.VisitAliasQualifiedName(node),  
        simplifier: SimplifyName);

    this.alwaysSimplify = oldAlwaysSimplify;

    return result;
}
```

to this:

``` C#
VisitAliasQualifiedName()
{
    return SimplifyVisit(node, n => SimplifyExpression(n, newNode: baseVisit(n), simplifier: SimplifyName)); 
}
```

### Known Issues
1.  We've discussed giving the user a dial in how and when these properties should be applied, e.g., "only apply this style on newly generated code" all the way up to "flag misuses of this style as a build break" and various steps in between.  That has not been addressed in this PR as it is still an open discussion item that I filed as #7289.
2.  If a user currently has "Qualify member access with '`this`' or '`Me`'" set to true, we need some way to upgrade/migrate that setting to the 4 new separate settings.

Fixes #7065 and #7090.

### CODE REVIEW

**Comment by brettfo** [comment]
test vsi please

**Comment by brettfo** [comment]
test prtest/win/dbg/unit32 please `// timeout trying to clone from GitHub`

p3 tests failed due to cross-cutting changes which are being tested [here](http://dotnet-ci.cloudapp.net/view/All/job/dotnet_ros-int_prtest_win_ccc/25/).  The corresponding internal PR is dotnet/roslyn-internal#430.

**Comment by brettfo** [comment]
Tagging @dotnet/roslyn-ide for review

**Comment by dpoeschl** [comment]
> 2. It is a well-known pattern that third-party code fixers should add the Simplifier.Annotation annotation to their generated code but they won't necessarily expand (adding this./Me.) before adding the annotation, and we want all code fixers (first- and third-party) to have the same nice experience of "I add the Simplifier.Annotation annotation and the user's code gets automatically corrected to their specified style."

Interesting. Are you proposing that simplify mean "apply *all* code style rules (as well as some other stuff that isn't code style rules but that you probably want simplified anyway)" going forward? This pattern probably makes sense for this code style rule, but it makes less sense for the naming convention code style rules (suddenly lots of renames are running when you only wanted to be removing unnecessary casts). How can we accommodate both use cases in a way that's comprehensible to people adding the `Simplifier.Annotation` (or do we indeed want to enforce even naming rules on Simplify)? Should the "other stuff that isn't code style rules" that we reduce become code style rules?

/cc: @Pilchie @DustinCampbell

**Comment by brettfo** [comment]
@dpoeschl: My thought was that "simplify" means "change anything that can't possibly be a breaking change" like whitespace, qualification, etc., but **not** naming because that could break API compat or reflection.

File: src/EditorFeatures/CSharpTest/CodeActions/EncapsulateField/EncapsulateFieldTests.cs [hunk 1]
```diff
@@ -1164,5 +1167,35 @@ class C
...
+            return this.i;
+        }
+        set
+        {
+            this.i = value;
+        }
+    }
+
+}
+", options: new Dictionary<OptionKey, object>() { { new OptionKey(SimplificationOptions.QualifyMemberFieldAccessWithThisOrMe, LanguageNames.CSharp), true } });
```
Comment (DustinCampbell) [inline src/EditorFeatures/CSharpTest/CodeActions/EncapsulateField/EncapsulateFieldTests.cs]: It'd be nice to have a test helper for constructing these.

File: src/Features/CSharp/Portable/CodeFixes/SimplifyTypeNames/SimplifyTypeNamesCodeFixProvider.cs [hunk 2]
```diff
@@ -28,7 +28,8 @@ public sealed override ImmutableArray<string> FixableDiagnosticIds
                 return ImmutableArray.Create(
                     IDEDiagnosticIds.SimplifyNamesDiagnosticId,
                     IDEDiagnosticIds.SimplifyMemberAccessDiagnosticId,
-                    IDEDiagnosticIds.SimplifyThisOrMeDiagnosticId);
+                    IDEDiagnosticIds.SimplifyThisOrMeDiagnosticId,
```
Comment (DustinCampbell) [inline src/Features/CSharp/Portable/CodeFixes/SimplifyTypeNames/SimplifyTypeNamesCodeFixProvider.cs]: Rename to `RemoveThisOrMeDiagnosticId` to be consistent?

File: src/Features/CSharp/Portable/CodeFixes/SimplifyTypeNames/SimplifyTypeNamesCodeFixProvider.cs [hunk 3]
```diff
@@ -96,6 +97,9 @@ internal static string GetCodeActionId(string diagnosticId, string nodeText)
                 case IDEDiagnosticIds.SimplifyThisOrMeDiagnosticId:
                     return CSharpFeaturesResources.SimplifyThisQualification;
```
Comment (DustinCampbell) [inline src/Features/CSharp/Portable/CodeFixes/SimplifyTypeNames/SimplifyTypeNamesCodeFixProvider.cs]: Rename to `RemoveThisQualification` to be consistent? (Note: that's what the resource actually is -- "Remove 'this' qualification")

File: src/VisualStudio/CSharp/Impl/CSharpVSResources.resx [hunk 4]
```diff
@@ -408,4 +408,16 @@
   <data name="NewLinesForBracesProperty" xml:space="preserve">
     <value>Place open brace on new line for properties, indexers, and events</value>
   </data>
+  <data name="QualifyMemberEventAccessWithThis" xml:space="preserve">
+    <value>Qualify member event access with 'this'</value>
```
Comment (DustinCampbell) [inline src/VisualStudio/CSharp/Impl/CSharpVSResources.resx]: I kind of feel like the words "member" and "event" should be swapped. Maybe "member could just go?

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs [hunk 5]
```diff
@@ -1114,10 +1126,67 @@ private static int GetNamespaceId(SyntaxList<MemberDeclarationSyntax> members, N
             return -1;
         }
 
+        private static bool TryNormalizeOrReduce(
```
Comment (DustinCampbell) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: What is meant by "Normalize"? This begs a descriptive comment.

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs [hunk 6]
```diff
@@ -1114,10 +1126,67 @@ private static int GetNamespaceId(SyntaxList<MemberDeclarationSyntax> members, N
...
+                    (symbol.IsKind(SymbolKind.Event) && optionSet.GetOption(SimplificationOptions.QualifyMemberEventAccessWithThisOrMe, semanticModel.Language)))
+                {
+                    replacementNode =
+                        SyntaxFactory.MemberAccessExpression(
+                            SyntaxKind.SimpleMemberAccessExpression,
+                            SyntaxFactory.ThisExpression(),
+                            (SimpleNameSyntax)name.WithLeadingTrivia().WithoutAnnotations(Simplifier.Annotation))
+                        .WithLeadingTrivia(name.GetLeadingTrivia());
+                    issueSpan = name.Span;
+                    return true;
```
Comment (DustinCampbell) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: I don't think this is the right thing to do here. The simplifier shouldn't introducing new nodes. It's job is to remove them. The way this should work is that the Expander should add `this.` in all cases (it already does) and the simplifier should remove them based on the current options.

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs (hunk 5 above)
Comment (DustinCampbell) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: Nevermind. See my comment below.

**Comment by DustinCampbell** [comment]
> I opted to make all changes in the reducer, including potentially expanding someField to this.someField/Me.someField. My reasons for doing this are twofold:
> 1. The reducer already has to handle the reduction or non-reduction of this./Me. and it seemed to make sense to do this work in the same location.
> 2. It is a well-known pattern that third-party code fixers should add the Simplifier.Annotation annotation to their generated code but they won't necessarily expand (adding this./Me.) before adding the annotation, and we want all code fixers (first- and third-party) to have the same nice experience of "I add the Simplifier.Annotation annotation and the user's code gets automatically corrected to their specified style."

Please don't expand nodes in the reducer. This is not how the system is supposed to work. Instead, simply reduce `this.` in the reducer based on the user's options. The expander already handles adding `this.` and it should be called first to add *all* `this.`s throughout the affected code. Then, the reducer can just work its magic.

**Comment by rchande** [comment]
Sounds like @DustinCampbell will also be a proponent of an entirely separate style-rule applying component.

**Comment by DustinCampbell** [comment]
@rchande, I don't have any context on what you mean.

**Comment by dpoeschl** [comment]
Just talked to @balajikris and @brettfo about the "style-rule applying component" for a while and landed mostly in agreement. Can bring you guys (@rchande & @DustinCampbell) up to speed tomorrow.

File: src/EditorFeatures/CSharpTest/CodeActions/GenerateFromMembers/GenerateConstructor/GenerateConstructorTests.cs [hunk 7]
```diff
@@ -93,7 +93,7 @@ protected override object CreateCodeRefactoringProvider(Workspace workspace)
             await TestAsync(
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/CodeActions/GenerateFromMembers/GenerateConstructor/GenerateConstructorTests.cs]: Add TestXXXAsync overloads that set up the options for you.

**Comment by CyrusNajmabadi** [comment]
Can you summarize the deletes in "src/EditorFeatures/Test2/Simplification/TypeNameSimplifierTest.vb ".  GitHub is not showing them.

File: src/VisualStudio/CSharp/Impl/CSharpVSResources.resx (hunk 4 above)
Comment (CyrusNajmabadi) [inline src/VisualStudio/CSharp/Impl/CSharpVSResources.resx]: I don't think "member" is needed in any of these. "Qualify event access with 'this'" reads much better to me.

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs [hunk 8]
```diff
@@ -1114,10 +1126,67 @@ private static int GetNamespaceId(SyntaxList<MemberDeclarationSyntax> members, N
...
 
+        private static bool TryNormalizeOrReduce(
+            this NameSyntax name,
+            SemanticModel semanticModel,
+            out ExpressionSyntax replacementNode,
+            out TextSpan issueSpan,
+            OptionSet optionSet,
+            CancellationToken cancellationToken)
+        {
+            if (name.TryNormalize(semanticModel, out replacementNode, out issueSpan, optionSet))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: return name.TryNormalize(...) || name.TryReduze(...)

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs [hunk 9]
```diff
@@ -699,10 +699,22 @@ private static bool CanReplace(ISymbol symbol)
...
             {
-                return false;
+                // if we can't resolve the symbol then nothing can be inferred
+                var nameSymbol = SimplificationHelpers.GetOriginalSymbolInfo(semanticModel, memberAccess.Name);
+                if (nameSymbol == null)
+                {
+                    return false;
+                }
+
+                if ((nameSymbol.IsKind(SymbolKind.Field) && optionSet.GetOption(SimplificationOptions.QualifyMemberFieldAccessWithThisOrMe, semanticModel.Language)) ||
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: This entire if statement seems to be repeated below.  Can you have a helper for this?

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs [hunk 10]
```diff
@@ -1114,10 +1126,67 @@ private static int GetNamespaceId(SyntaxList<MemberDeclarationSyntax> members, N
...
+            var symbol = SimplificationHelpers.GetOriginalSymbolInfo(semanticModel, name);
+            if (symbol == null)
+            {
+                return false;
+            }
+
+            // if not static and not already part of a `this.` expression
+            if (!symbol.IsStatic &&
+                name is SimpleNameSyntax &&
+                (name.Parent as MemberAccessExpressionSyntax)?.Expression.Kind() != SyntaxKind.ThisExpression)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: What about named on the right of things like ```->``` or ```?.```  I think we have syntax helpers that can ask if this is the right side of a some sort of qualified name or not.  Can you use those instead?

File: src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs [hunk 11]
```diff
@@ -1114,10 +1126,67 @@ private static int GetNamespaceId(SyntaxList<MemberDeclarationSyntax> members, N
...
+                    (symbol.IsKind(SymbolKind.Property) && optionSet.GetOption(SimplificationOptions.QualifyMemberPropertyAccessWithThisOrMe, semanticModel.Language)) ||
+                    (symbol.IsKind(SymbolKind.Method) && optionSet.GetOption(SimplificationOptions.QualifyMemberMethodAccessWithThisOrMe, semanticModel.Language)) ||
+                    (symbol.IsKind(SymbolKind.Event) && optionSet.GetOption(SimplificationOptions.QualifyMemberEventAccessWithThisOrMe, semanticModel.Language)))
+                {
+                    replacementNode =
+                        SyntaxFactory.MemberAccessExpression(
+                            SyntaxKind.SimpleMemberAccessExpression,
+                            SyntaxFactory.ThisExpression(),
+                            (SimpleNameSyntax)name.WithLeadingTrivia().WithoutAnnotations(Simplifier.Annotation))
+                        .WithLeadingTrivia(name.GetLeadingTrivia());
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Extensions/ExpressionSyntaxExtensions.cs]: What about trailing trivia?

File: src/Workspaces/CSharp/Portable/Simplification/CSharpNameReducer.Rewriter.cs [hunk 12]
```diff
@@ -19,160 +19,72 @@ public Rewriter(OptionSet optionSet, CancellationToken cancellationToken)
             {
             }
 
-            public override SyntaxNode VisitPredefinedType(PredefinedTypeSyntax node)
+            private SyntaxNode SimplifyVisit<TExpression>(
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Simplification/CSharpNameReducer.Rewriter.cs]: I appreciate the simplification.  However, I worry about all the delegate creations you're now causing.  can you capture the delegates in the constructor and then pass them in accordingly?

Remember, we're going to walk huge trees.  Causing a lambda allocation on every identifier (for example) is not good.

File: src/Workspaces/Core/Portable/PublicAPI.Unshipped.txt [hunk 13]
```diff
@@ -4,4 +4,8 @@ Microsoft.CodeAnalysis.Workspace.UpdateReferencesAfterAdd() -> void
 abstract Microsoft.CodeAnalysis.Editing.SyntaxGenerator.ArrayCreationExpression(Microsoft.CodeAnalysis.SyntaxNode elementType, Microsoft.CodeAnalysis.SyntaxNode size) -> Microsoft.CodeAnalysis.SyntaxNode
 abstract Microsoft.CodeAnalysis.Editing.SyntaxGenerator.ArrayCreationExpression(Microsoft.CodeAnalysis.SyntaxNode elementType, System.Collections.Generic.IEnumerable<Microsoft.CodeAnalysis.SyntaxNode> elements) -> Microsoft.CodeAnalysis.SyntaxNode
 static Microsoft.CodeAnalysis.Editing.SyntaxGenerator.GetGenerator(Microsoft.CodeAnalysis.Project project) -> Microsoft.CodeAnalysis.Editing.SyntaxGenerator
+static Microsoft.CodeAnalysis.Simplification.SimplificationOptions.QualifyMemberEventAccessWithThisOrMe.get -> Microsoft.CodeAnalysis.Options.PerLanguageOption<bool>
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/PublicAPI.Unshipped.txt]: So should this just be "QualifyMemberAccess"?  And should this feature also control qualifying static accesses?

File: src/Workspaces/Core/Portable/Simplification/SimplificationOption.cs [hunk 14]
```diff
@@ -45,9 +46,30 @@ public static class SimplificationOptions
         public static Option<bool> AllowSimplificationToBaseType { get; } = new Option<bool>(NonPerLanguageFeatureName, "AllowSimplificationToBaseType", true);
 
         /// <summary>
-        /// This option says if we should simplify away the this. or Me. in member access expression
+        /// This option says if we should simplify away the <see langword="this"/>. or <see langword="Me"/>. in member access expressions.
         /// </summary>
-        public static PerLanguageOption<bool> QualifyMemberAccessWithThisOrMe { get; } = new PerLanguageOption<bool>(PerLanguageFeatureName, "QualifyMemberAccessWithThisOrMe", defaultValue: false);
+        [Obsolete]
+        public static PerLanguageOption<bool> QualifyMemberAccessWithThisOrMe { get; } = new PerLanguageOption<bool>(PerLanguageFeatureName, "QualifyMemberAccessWithThisOrMe", false);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/Simplification/SimplificationOption.cs]: I really hate seeing "ThisOrMe".  It's essentially putting VB and C#isms into a language neutral layer.  Can we instead just have this be QualifyInstanceMemberAccess and QualifyStaticMemberAccess?

File: src/Workspaces/VisualBasic/Portable/Extensions/ExpressionSyntaxExtensions.vb [hunk 15]
```diff
@@ -949,9 +949,20 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Extensions
...
-                 memberAccess.Expression.Kind() = SyntaxKind.MeExpression Then
-                Return False
+            If memberAccess.Expression.IsKind(SyntaxKind.MeExpression) Then
+                ' if we can't resolve the symbol then nothing can be inferred
+                Dim nameSymbol = SimplificationHelpers.GetOriginalSymbolInfo(semanticModel, memberAccess.Name)
+                If nameSymbol Is Nothing Then
+                    Return False
+                End If
+
+                If (nameSymbol.IsKind(SymbolKind.Field) AndAlso optionSet.GetOption(SimplificationOptions.QualifyMemberFieldAccessWithThisOrMe, semanticModel.Language)) OrElse
```
Comment (CyrusNajmabadi) [inline src/Workspaces/VisualBasic/Portable/Extensions/ExpressionSyntaxExtensions.vb]: This code looks familiar.

File: src/Workspaces/VisualBasic/Portable/Extensions/ExpressionSyntaxExtensions.vb [hunk 16]
```diff
@@ -1151,6 +1162,64 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Extensions
...
+
+            If name.TryNormalize(semanticModel, replacementNode, issueSpan, optionSet) Then
+                Return True
+            End If
+
+            Return name.TryReduce(semanticModel, replacementNode, issueSpan, optionSet, cancellationToken)
+        End Function
+
+        <Extension>
+        Private Function TryNormalize(
```
Comment (CyrusNajmabadi) [inline src/Workspaces/VisualBasic/Portable/Extensions/ExpressionSyntaxExtensions.vb]: As does this.  Why can't we share most of this code between VB and C#?

File: src/Workspaces/VisualBasic/Portable/Simplification/VisualBasicNameReducer.Rewriter.vb [hunk 17]
```diff
@@ -14,100 +14,51 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Simplification
                 MyBase.New(optionSet, cancellationToken)
             End Sub
 
-            Public Overrides Function VisitGenericName(node As GenericNameSyntax) As SyntaxNode
+            Private Function SimplifyVisit(Of TExpression As SyntaxNode)(
```
Comment (CyrusNajmabadi) [inline src/Workspaces/VisualBasic/Portable/Simplification/VisualBasicNameReducer.Rewriter.vb]: Same concern about lambda allocations.  Please cache the delegates in the constructor and use the cached values in the code.

**Comment by CyrusNajmabadi** [comment]
I agree with dustin.  The simplifier should not be expanding nodes.  The complexifier/expander should do this, and the simplifier can decide if it should be simplifying or not.

File: src/Workspaces/CSharp/Portable/Simplification/CSharpNameReducer.Rewriter.cs (hunk 12 above)
Comment (DustinCampbell) [inline src/Workspaces/CSharp/Portable/Simplification/CSharpNameReducer.Rewriter.cs]: Great point -- I hadn't considered that.

**Comment by brettfo** [comment]
Closing in favor of #7540.
